### DISCUSSION TITLE
AI deck control LED Ring state

### DISCUSSION THREAD

**Discussion by LucasTheIntern** [bitcraze/discussions#905] [General]
Hi! 

I'm working on a project that requires the AI deck to perform on-board image processing, and then alter the state of the LED Ring deck (basic example: turn the LED ring on/off depending on brightness of camera frame). 

Everything that I want the Crazyflie to accomplish is easily possible by streaming images to a laptop and then having a script do the processing and communicate with the drone... but having everything onboard is a major part of my project. 

I will also have the Lighthouse deck on board for navigation/geofencing. I am aware that the Lighthouse deck and the AI deck have conflicts if certain pins on the AI deck are enabled, but I think I can avoid this... My plan is to mount the AI deck between the main deck and the LED ring (attached image), and only connect the AI deck to the main deck via the pins:
SDA
SCL
BOOT
RST
VCOM

per the specifications in the compatibility matrix. None of these appear to have conflicts with the Lighthouse deck.

**Question #1:**
Can I get the AI deck to write PWM signals to the IO2 and IO3 pins? I assume these are the pins used to change the state of the LED ring by the main deck. If not, could I use some of the RX/TX pins instead and make a wired connection to the IO pins on the LED ring? Other suggestions/pointers for getting this to work are welcome. 

**Question #2:**
If it is indeed possible to write PWM signals to the LED Ring, where can I find out what signals I should send (without going to the trouble of measuring them)? I have not yet found documentation on this.

Thank you!
Lucas

![image6](https://github.com/bitcraze/discussions/assets/120214901/dd5fe6fd-b28d-468a-87f8-187ee5180285)

**Comment by krichardsson** [comment]
The way to do this is to create an app in the Crazyflie STM that controls the LED ring. Use [CPX](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/) to send messages from the AI-deck (ESP or GAP) to your app.

You can find an example app that communicates with the GAP8 in the [app_stm_gap8_cpx app](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/app_stm_gap8_cpx) together with the [stm_gap8_cpx GAP8 example](https://github.com/bitcraze/aideck-gap8-examples/tree/master/examples/other/stm_gap8_cpx). The STM part will also show how to make an app.

To control the LED ring, use parameters. You can try it out from the python client by setting parameters in the [`ring` group](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#ring), perhaps setting [`ring.effect`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#ringeffect) to 7 and [`ring.solidRed`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#ringsolidred), [`ring.solidGreen`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#ringsolidgreen), [`ring.solidBlue`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#ringsolidblue) to your favorite color.
When you know what values you want for your parameters you will have to implement this in your app, you can find an example of how to set parameters from an app in [app_internal_param_log](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/app_internal_param_log)
